source/ConditionSet.cpp

```diff
@@ -325,477 +361,372 @@ set<string> ConditionSet::RelevantConditions() const
-	while(!UsedAll(usedOps))
-	{
-		while(true)
-		{
-			// Stack ops until one of lower or equal precedence is found, then evaluate the higher one first.
-			if(opStack.empty() || operators.at(opIndex) == "("
-					|| (Precedence(operators.at(opIndex)) > Precedence(operators.at(opStack.back()))))
-			{
-				opStack.push_back(opIndex);
-				// Mark this operator as used and advance.
-				usedOps.at(opIndex++) = true;
-				break;
-			}
-
-			size_t workingIndex = opStack.back();
-			opStack.pop_back();
-
-			// A left parentheses results in a no-op step.
-			if(operators.at(workingIndex) == "(")
-			{
-				if(operators.at(opIndex) != ")")
-				{
-					Logger::LogError("Did not find matched parentheses:");
-					PrintConditionError(ToStrings());
-					tokens.clear();
-					operators.clear();
-					sequence.clear();
-					return;
-				}
-				// "Use" the parentheses and advance operators.
-				usedOps.at(opIndex++) = true;
-				break;
-			}
-			else if(!AddOperation(dataDest, destinationIndex, workingIndex))
-				return;
-		}
-	}
-	// Handle remaining operators (which cannot be parentheses).
-	while(!opStack.empty())
-	{
-		size_t workingIndex = opStack.back();
-		opStack.pop_back();
+	ConditionSet ce(*this);
+	children.clear();
+	children.emplace_back(ce);
+	expressionOperator = OP_AND;
 
-		if(operators.at(workingIndex) == "(" || operators.at(workingIndex) == ")")
-		{
-			Logger::LogError("Mismatched parentheses:" + ToString());
-			tokens.clear();
-			operators.clear();
-			sequence.clear();
-			return;
-		}
-		else if(!AddOperation(dataDest, destinationIndex, workingIndex))
-			return;
-	}
-	// All operators and tokens should now have been used.
+	return true;
 }
 
 
 
-// Use a valid working index and data pointer vector to create an evaluable Operation.
-bool ConditionSet::Expression::SubExpression::AddOperation(vector<int> &data, size_t &index, const size_t &opIndex)
+bool ConditionSet::PushDownLast(const DataNode &node)
 {
-	// Obtain the operand indices. The operator is never a parentheses. The
-	// operator index never exceeds the size of the tokens vector.
-	size_t leftIndex = FindOperandIndex(tokens, data, opIndex, true);
-	size_t rightIndex = FindOperandIndex(tokens, data, opIndex, false);
-
-	// Bail out if the pointed token is in-bounds and empty.
-	if((leftIndex < tokens.size() && tokens.at(leftIndex).empty())
-			|| (rightIndex < tokens.size() && tokens.at(rightIndex).empty()))
-	{
-		Logger::LogError("Unable to obtain valid operand for function \"" + operators.at(opIndex) + "\" with tokens:");
-		PrintConditionError(tokens);
-		tokens.clear();
-		operators.clear();
-		sequence.clear();
-		return false;
-	}
+	// Can only perform push-down if there is at least one expression to push down.
+	if(children.empty())
+		return FailParse(node, "cannot create sub-expression from void");
+
+	ConditionSet ce = children.back();
+	children.pop_back();
+	children.emplace_back();
+	children.back().children.push_back(std::move(ce));
```

Comment:
You used emplace_back in PushDownFull, is one of the options better?

Reply:
`push_back()` feels  semantically better than `emplace_back()`, since we already have the object that we want to push (and thus don't need copy or move constructors).
I expect that in practice the compiler generates the same code for both of them.

Reply:
Updated both to push_back.